pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1214,4 +1273,112 @@ private int getMaxNumPartitionsPerInstance(InstancePartitions instancePartitions
       return (numPartitions + numInstancesPerReplicaGroup - 1) / numInstancesPerReplicaGroup;
     }
   }
+
+  // Pre fetch the LLC segment without deep store copy.
+  public void prefetchLLCSegmentWithoutDeepStoreCopy(String tableNameWithType) {
+      TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableNameWithType);
+      if (tableType == TableType.REALTIME) {
+        TableConfig tableConfig = _helixResourceManager.getTableConfig(tableNameWithType);
+        if (tableConfig == null) {
+          LOGGER.warn("Failed to find table config for table: {}", tableNameWithType);
+          return;
+        }
+
+        PartitionLevelStreamConfig streamConfig = new PartitionLevelStreamConfig(tableConfig.getTableName(),
+            IngestionConfigUtils.getStreamConfigMap(tableConfig));
+        if (!streamConfig.hasLowLevelConsumerType()) {
+          return;
+        }
+
+        List<LLCRealtimeSegmentZKMetadata> segmentZKMetadataList = ZKMetadataProvider.getLLCRealtimeSegmentZKMetadataListForTable(_propertyStore, tableNameWithType);
```

Comment:
this is an very expensive operation for a large table with thousands of segments? can we reduce the impact of such call?

Reply:
one option is to first get the list of segments; do some filtering on it (based on segment creation time) and then only fetch zkmeta data for the filtered segments.

Reply:
I believe I had suggested an alternative here.

When controller comes up (or gains mastership for a table), scan all tables for segments missing ini deep store. Keep the list of missing segments in memory.

Whenever job is triggered, pull missing segment list from memory. If the upload succeeds, delete the segment from memory list.

Whenever a deepstore upload fails (recognized during commit), add segment to in-memory list.

Reply:
Hi @mcvsubbu , this change is following your suggestion. Please correct me if there is any missing steps.

During commit phase, [enqueue the segment without download url. ](https://github.com/apache/incubator-pinot/pull/6778/files/647f1de17ff0bc60bef20468604e65b44dbc057c#diff-ffc91e381e1c93dc73fbeb18ef872ec812e25f15760dbf35896ac81793ce1795R643)

`uploadToSegmentStoreIfMissing`  reads in-memory segments list to fix, only removes the segment from the in-memory list after successful upload.

Only `prefetchLLCSegmentWithoutDeepStoreCopy ` has access to ZK, [which is triggered when setup the periodic jobs](https://github.com/apache/incubator-pinot/pull/6778/files/647f1de17ff0bc60bef20468604e65b44dbc057c#diff-0a9ffdab99ed314e7a8e58bf86c125f020b7d6f86493d1f6d719beab62b7f106R71). Leadership is also checked in this step.

Reply:
@chenboat Added the filter based on creation time from LLC segment name.